src/Microsoft.AspNetCore.Mvc.Core/Internal/MvcEndpointInvokerFactory.cs

```diff
@@ -0,0 +1,41 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using Microsoft.AspNetCore.Mvc.Abstractions;
+using Microsoft.AspNetCore.Mvc.Infrastructure;
+
+namespace Microsoft.AspNetCore.Mvc.Internal
+{
+    internal sealed class MvcEndpointInvokerFactory : IActionInvokerFactory
+    {
+        private readonly IActionInvokerFactory _invokerFactory;
+        private readonly IActionContextAccessor _actionContextAccessor;
+
+        public MvcEndpointInvokerFactory(
+            IActionInvokerFactory invokerFactory)
+            : this(invokerFactory, actionContextAccessor: null)
+        {
+        }
+
+        public MvcEndpointInvokerFactory(
+            IActionInvokerFactory invokerFactory,
+            IActionContextAccessor actionContextAccessor)
+        {
+            _invokerFactory = invokerFactory;
+
+            // The IActionContextAccessor is optional. We want to avoid the overhead of using CallContext
+            // if possible.
+            _actionContextAccessor = actionContextAccessor;
+        }
+
+        public IActionInvoker CreateInvoker(ActionContext actionContext)
+        {
+            if (_actionContextAccessor != null)
+            {
+                _actionContextAccessor.ActionContext = actionContext;
+            }
```

Comment:
This should be inside another delegate that wraps the result of `CreateInvoker`. It's not really the current context until you start executing it.

Reply:
I lifted this logic from https://github.com/aspnet/Mvc/blob/dev/src/Microsoft.AspNetCore.Mvc.Core/Internal/MvcAttributeRouteHandler.cs#L92-L111

Also the dispatching functional tests started working once I added this code to set the context.

Reply:
OK checking into this a bit more, this works because you call the invoker factory inside the route handler. This seems OK for now, but this could regress if we refactor some other stuff in fairly innocent ways.

Reply:
There are functional tests that rely on it being set, but I'll add a unit test for it as well.